File: backend/app/schemas/rule.py
"""
规则引擎 schema — RuleDef + CRUD 入参/出参

RuleDef: 规则的内部表示，logic 字段存 json-logic 对象
RuleCreate / RuleResponse: 规则管理 API 的入参/出参
"""
from typing import Any, Dict, List, Literal, Optional
from pydantic import BaseModel, Field


# ========== 规则执行模式 ==========

RuleExecMode = Literal["deterministic", "pre_computed", "semantic"]
"""
- deterministic: json-logic 直接求值, 零次 LLM 调用
- pre_computed: build_data 预计算后再 json-logic 求值, 零次 LLM 调用
- semantic:      LLM 语义判断 (招待费备注是否合理等)"
"""


# ========== 规则内部表示 (给 RuleEngine 用) ==========

class RuleDef(BaseModel):
    """一条可被 RuleEngine 求值的规则"""
    name: str = Field(..., description="规则名称")
    rule_type: str = Field(..., description="规则类型 (amount_limit/compliance/time/…)")
    logic: Dict[str, Any] = Field(
        ...,
        description="json-logic 对象, 真正被求值的部分。例: {'<=':[{'var':'total_amount'},5000]}"
    )
    action: Literal["reject", "warn", "require_approval"] = Field(
        ..., description="规则命中后的动作"
    )
    message: str = Field(..., description="规则命中后的提示文案")
    description: Optional[str] = Field(None, description="人类可读规则说明")
    exec_mode: RuleExecMode = Field(
        default="deterministic",
        description="执行模式: deterministic / pre_computed / semantic"
    )
    is_active: bool = Field(default=True, description="是否启用")

    class Config:
        # 允许直接传 dict 构造（方便从 DB/文件读取后转换）
        extra = "ignore"


# ========== CRUD Schema ==========

class RuleCreate(BaseModel):
    """创建规则请求"""
    name: str = Field(..., min_length=1, max_length=200, description="规则名称")
    rule_type: str = Field(
        ..., description="规则类型, 对应 RuleType 枚举值"
    )
    logic: Dict[str, Any] = Field(
        ..., description="json-logic 规则对象"
    )
    action: Literal["reject", "warn", "require_approval"] = Field(
        ..., description="规则动作"
    )
    message: str = Field(..., min_length=1, max_length=500, description="命中提示")
    description: Optional[str] = Field(None, max_length=500, description="人类可读说明")
    exec_mode: RuleExecMode = Field(
        default="deterministic", description="执行模式"
    )


class RuleUpdate(BaseModel):
    """更新规则请求 — 所有字段可选"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    rule_type: Optional[str] = None
    logic: Optional[Dict[str, Any]] = None
    action: Optional[Literal["reject", "warn", "require_approval"]] = None
    message: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = Field(None, max_length=500)
    exec_mode: Optional[RuleExecMode] = None
    is_active: Optional[bool] = None


class RuleResponse(BaseModel):
    """规则查询响应"""
    id: int
    name: str
    rule_type: str
    logic: Dict[str, Any]
    action: str
    message: str
    description: Optional[str] = None
    exec_mode: RuleExecMode = "deterministic"
    is_active: bool = True

    class Config:
        from_attributes = True


class RuleListResponse(BaseModel):
    """规则列表响应"""
    total: int
    items: List[RuleResponse]

File: frontend/src/types/rule.ts
/** 后端 RuleType 八值 */
export const RuleTypeLabels: Record<string, string> = {
  amount_limit: '金额限制', frequency: '频率限制', category: '类别限制',
  department: '部门限制',   position: '职位限制',  time: '时间限制',
  compliance: '合规规则',   custom: '自定义规则'
}

export const RuleActionLabels: Record<string, string> = {
  reject: '拒绝', warn: '警告', require_approval: '需额外审批'
}
export const RuleActionColors: Record<string, string> = {
  reject: 'danger', warn: 'warning', require_approval: 'info'
}

export const RuleExecModeLabels: Record<string, string> = {
  deterministic: '确定性求值', pre_computed: '预计算求值', semantic: '语义(LLM)判断'
}

export type RuleAction   = 'reject' | 'warn' | 'require_approval'
export type RuleExecMode = 'deterministic' | 'pre_computed' | 'semantic'

/** 严格对齐后端 RuleResponse —— 注意：没有 created_at */
export interface Rule {
  id: number
  name: string
  rule_type: string
  logic: Record<string, any>      // 存量数据可能是 {}
  action: RuleAction
  message: string                 // 后端伪字段，恒为 `${name}不符合规则`
  description?: string | null     // 实际映射 ORM 的 condition 列
  exec_mode: RuleExecMode
  is_active: boolean
}

export interface RuleListResponse { total: number; items: Rule[] }

export interface RuleCreatePayload {
  name: string
  rule_type: string
  logic: Record<string, any>      // 必填且必须非空
  action: RuleAction
  message: string                 // 必填（否则 422），但后端不持久化
  description?: string
  exec_mode?: RuleExecMode
}

/** 全部可选；未传的字段后端不会动 —— logic 省略即跳过校验 */
export type RuleUpdatePayload = Partial<RuleCreatePayload> & { is_active?: boolean }

export interface RuleListParams {
  page?: number; page_size?: number
  rule_type?: string; is_active?: boolean; exec_mode?: string
}

/** 与后端 rule_engine.FIELD_WHITELIST 完全一致（用于前端预校验） */
export const RULE_FIELD_WHITELIST = [
  'total_amount', 'currency', 'title', 'expense_type', 'description', 'items',
  'user_department', 'user_role',
  'item_count', 'has_unverified_invoice', 'max_invoice_age_days'
] as const

/** 与后端 rule_engine.ALLOWED_OPS 完全一致 */
export const RULE_ALLOWED_OPS = [
  '==','===','!=','!==','>','>=','<','<=',
  '!','!!','and','or','?:','if',
  'var','missing','missing_some',
  'map','reduce',
  '%','+','*','-','/','min','max',
  'cat','substr','in','merge','count',
  'today','date','datetime','rdelta','duration','log'
] as const
